[IMP] mail: Improve discuss (call) actions + meeting view

alexkuhn opened the pull request:
1. Call actions now use ActionList, further harmonizing definition and style of discuss actions.
2. ActionList have generic `more()` action to nest actions in a dropdown, following definition and style of action. This can be used wherever the business code needs in the partition of actions.
3. Split call actions in layout and non-layout ones. The layout actions (pip and fullscreen) are now available in bottom-right of call view ("Layout actions") instead of call action dropdown.
4. improve meeting view action panels: this now uses resizable panels like Discuss app, and many more actions are available. Also fixes issue with message highlight to work with chat panel and also jump to other thread than the meeting chat.
5. improve thread / message / composer / call action definitions: simpler syntax, harmonize them further, more patchable features, JS doc typing. This eases reading and writing discuss actions.

https://github.com/odoo/enterprise/pull/93712

tsm-odoo commented on addons/im_livechat/static/src/core/web/thread_actions.js:
```diff
@@ -5,49 +5,27 @@ import { LivechatChannelInfoList } from "@im_livechat/core/web/livechat_channel_
-    },
+    condition: ({ owner, thread }) =>
+        thread?.channel_type === "livechat" && !thread.livechat_end_dt && !owner.isDiscussContent,
+    dropdown: 1,
```
I'd rather having a boolean. Name doesn't suggest a boolean, I was wondering if it was a sequence, or something else?

tsm-odoo commented on addons/mail/static/src/core/common/chat_hub.js:
```diff
@@ -67,26 +68,44 @@ export class ChatHub extends Component {
-                icon: "oi oi-close",
-                onSelected: () => this.chatHub.closeAll(),
-            });
+            actions.push(
+                new Action({
+                    owner: this,
+                    id: "hide-all",
+                    definition: {
+                        name: _t("Hide all conversations"),
+                        icon: "fa fa-eye-slash",
+                        onSelected: () => this.chatHub.hideAll(),
+                    },
+                    store: this.store,
+                })
+            );
+            actions.push(
+                new Action({
+                    owner: this,
+                    id: "close-all",
+                    definition: {
+                        name: _t("Close all conversations"),
+                        icon: "oi oi-close",
+                        onSelected: () => this.chatHub.closeAll(),
+                    },
+                    store: this.store,
+                })
+            );
         }
```
:pick: 

```suggestion
        actions.push(
            new Action({
                owner: this,
                id: "hide-all",
                definition: {
                    name: _t("Hide all conversations"),
                    icon: "fa fa-eye-slash",
                    onSelected: () => this.chatHub.hideAll(),
                },
                store: this.store,
            }),
            new Action({
                owner: this,
                id: "close-all",
                definition: {
                    name: _t("Close all conversations"),
                    icon: "oi oi-close",
                    onSelected: () => this.chatHub.closeAll(),
                },
                store: this.store,
            })
        );
```

tsm-odoo reviewed (commented):
Some comments

alexkuhn commented on addons/crm_livechat/static/src/core/thread_actions.js:
```diff
@@ -8,34 +8,33 @@ import { usePopover } from "@web/core/popover/popover_hook";
     sequenceGroup: 25,
-    setup(component) {
-        if (!component.env.inChatWindow) {
+    setup({ owner }) {
```
Indeed, the PR changes many owner accessors into something else. The play forward is replace most (if not all) `owner` occurrences.

alexkuhn commented:
> * In the meeting view, it's very disturbing that panels have different size. [...] I'd rather keeping a base size for the meeting view panel.

Fixed

> * There's a small position conflict with meeting view panels

Will look at it in follow-up PR, this needs to tackle arrangeTiles() in Call view which is easier said than done :D

> * Invite ppl/Sub channel dropdown are too small now IMO.

Fixed

> * Small alignment issue with some buttons having rounded-circle+background

Unfortunately a problem with the way circles are done in dropdown, using a `i:before`, which depends on how font is rendered and differs between browsers and OS. Already present in master, will need its own fix eventually.

tsm-odoo commented on addons/mail/static/src/core/common/chat_hub.js:
```diff
@@ -67,26 +68,42 @@ export class ChatHub extends Component {
+                        icon: "fa fa-undo",
+                        onSelected: () => this.resetPosition(),
+                    },
+                    store: this.store,
```
thats a bit cumbersome to pass store every single time while its a singleton.

tsm-odoo commented on addons/mail/static/src/core/common/composer_actions.js:
```diff
@@ -120,103 +111,79 @@ registerComposerAction("add-emoji", {
+    get params() {
+        const res = super.params;
+        const self = this;
+        Object.defineProperty(res, "composer", {
+            get() {
+                return self.composerFn();
+            },
+            set: () => {}, // avoids Proxy "trap returned falsish" error
+        });
```
Can probably be removed?

tsm-odoo commented on addons/mail/static/src/core/common/composer_model.js:
```diff
@@ -94,6 +94,10 @@ export class Composer extends Record {
     }
+
+    get targetThread() {
+        return this.replyToMessage?.thread ?? this.thread ?? null;
```
do we really need the fallback on null?

alexkuhn replied:
I don't think so, this is a move from component to record for the sake of avoiding a `owner.thread`.
To make sure to not break things in last minute, let's keep it there for now.

tsm-odoo commented on addons/mail/static/src/core/common/message_actions.js:
```diff
@@ -3,31 +3,29 @@ import { toRaw, useComponent, useState } from "@odoo/owl";
 import { QuickReactionMenu } from "@mail/core/common/quick_reaction_menu";
 import { isMobileOS } from "@web/core/browser/feature_detection";
-import { Action, UseActions } from "./action";
+import { Action, ACTION_TAGS, UseActions } from "./action";
```
:pick: use full paths :)

tsm-odoo commented on addons/mail/static/src/core/public_web/discuss_search.xml:
```diff
@@ -12,7 +12,7 @@
         <t t-else="" t-call="mail.DiscussSearch.inputContainer"/>
         <t t-if="store.self.main_user_id">
-            <Dropdown t-if="store.discuss.isSidebarCompact and !ui.isSmall" state="meetingFloating" position="'right-start'" menuClass="'o-mail-DiscussSidebar-floatingMenu o-discuss-dropdownMenu border-secondary px-2 mx-2 my-0 min-w-0'" manual="true">
+            <Dropdown t-if="store.discuss.isSidebarCompact and !ui.isSmall" state="meetingFloating" position="'right-start'" menuClass="'o-mail-DiscussSidebar-floatingMenupx-2 mx-2 my-0 min-w-0 ' + store.discussDropdownMenuClass(this)" manual="true">
```
```suggestion
            <Dropdown t-if="store.discuss.isSidebarCompact and !ui.isSmall" state="meetingFloating" position="'right-start'" menuClass="'o-mail-DiscussSidebar-floatingMenu px-2 mx-2 my-0 min-w-0 ' + store.discussDropdownMenuClass(this)" manual="true">
```

tsm-odoo commented on addons/mail/static/src/discuss/call/common/call_action_list.js:
```diff
@@ -20,12 +21,49 @@ export class CallActionList extends Component {
+                    (a) => !a.tags.includes(ACTION_TAGS.CALL_LAYOUT)
+                );
+                const sequenceGroup = filtered[0].sequenceGroup;
+                // const maxQuickActions = this.env.inChatWindow && sequenceGroup === 200 ? 0 : 4;
```
:see_no_evil:

tsm-odoo commented on addons/mail/static/src/discuss/call/common/meeting.js:
```diff
@@ -3,54 +3,68 @@ import { Thread } from "@mail/core/common/thread";
-        this.state = useState({
-            activeSidePanel: this.props.initialSidePanel,
-            jumpPresent: 0,
+        onMounted(() => {
+            if (this.props.autoOpenAction) {
+                this.threadActions.actions
+                    .find((a) => a.id === this.props.autoOpenAction)
+                    ?.onSelected();
+            }
+        });
```
This is much more fragile than the state that was previously used. Component could mount for several reasons, changing the active action.

tsm-odoo commented on addons/mail/static/src/discuss/call/common/meeting_chat.js:
```diff
@@ -0,0 +1,36 @@
+        Composer,
+        Thread,
+    };
+    static props = ["*"];
```
It's sad not to benefit from validation.

alexkuhn commented on addons/web/static/src/core/dropzone/dropzone.js:
```diff
@@ -4,7 +4,7 @@ export class Dropzone extends Component {
         extraClass: { type: String, optional: true },
         onDrop: { type: Function, optional: true },
-        ref: Object,
+        ref: [Object, Function],
```
Just props validation that was not taking `useForwardRefToParent` into account.

alexkuhn commented:
Would have been best to make changes after https://github.com/odoo/enterprise/pull/89292 but its ci/upgrade failed
@robodoo r+

robodoo commented:
@alexkuhn staging failed: ci/runbot (view more at https://runbot.odoo.com/runbot/batch/2122224/build/88517461)

alexkuhn commented:
Usage of new `action.more()` feature is prone to double rendering that runbot error shows.

The `.more()` feature could have better design to significantly reduce risks of that, but this is hard to do without making the feature even more cumbersome or more limiting functionally than now. The `.more()` inside `onWillRender()` allows lots of flexibility for component needs, decide what are conditions for having more, which group(s) are affected, at which position, etc.

Hopefully can find a way to have equivalent flexibility while making it less prone to double rendering. For now I've put some `toRaw()` in the `onWillRender` in attempt to prevent double rendering while still ensuring enough reactive observe on the `more()` feature (i.e. it should open/close dropdown reactively + re-render quick message action list on close / click away).

alexkuhn commented:
Extra renderings that were highlighted by `discuss_performance.test.js` come from using `onWillRender` in `Message` component whose whole template is inside slotted component `ActionSwiper`.

Because actions have reactives (data related to JS models with lazy computed, internal state like dropdowns), they trigger re-render with notifyReactives which is fine... But with `ActionSwiper` layer this has time to re-render without the slotted template, and thus it renders twice from lazy computes...

I added a commit to replace `onWillRender` by an equivalent `computeActions` just below `ActionSwiper`, so that we are not affected by the `ActionSwiper` layer.

Hopefully this fixes the issue on runbot with `discuss_performance.test.js`. Other users of `Action.more()` have not their whole template in slotted component like `Message` so they were unaffected by the issue.

alexkuhn commented:
@robodoo r+

robodoo commented:
@alexkuhn because this PR has multiple commits, I need to know how to merge it:

* `merge` to merge directly, using the PR as merge commit message
* `rebase-merge` to rebase and merge, using the PR as merge commit message
* `rebase-ff` to rebase and fast-forward

alexkuhn commented:
@robodoo rebase-ff

robodoo commented:
Merge method set to rebase and fast-forward.
